File: examples/class-kits/gospider/ecommerce_detail.go
package classkit

import scrapyapi "gospider/scrapy"

func NewEcommerceDetailSpider() *scrapyapi.Spider {
	profile := profileForFamily(defaultSiteFamily)
	return scrapyapi.NewSpider("ecommerce-detail", func(response *scrapyapi.Response) ([]any, error) {
		family := siteFamilyFromResponse(response.URL, nil)
		if response.Request != nil {
			family = siteFamilyFromResponse(response.URL, response.Request.Meta)
		}
		current := profileForFamily(family)
		title := bestTitle(response.CSS("title").Get(), response.CSS("h1").Get())
		links := response.XPath("//a/@href").GetAll()

		return []any{
			scrapyapi.NewItem().
				Set("kind", "ecommerce_detail").
				Set("site_family", family).
				Set("title", title).
				Set("url", response.URL).
				Set("item_id", firstRegexMatch(response.Text, current.ItemIDPatterns)).
				Set("price", firstRegexMatch(response.Text, current.PricePatterns)).
				Set("shop", firstRegexMatch(response.Text, current.ShopPatterns)).
				Set("review_count", firstRegexMatch(response.Text, current.ReviewCountPatterns)).
				Set("image_candidates", collectImageLinks(response.URL, response.XPath("//img/@src").GetAll(), 10)).
				Set("review_url", firstLinkWithKeywords(response.URL, links, current.ReviewLinkKeywords)).
				Set("html_excerpt", excerpt(response.Text, 800)).
				Set("note", "Template for public product detail pages. Extend with site-specific JSON/bootstrap extraction when available."),
		}, nil
	}).
		AddStartURL(profile.DetailURL).
		WithStartMeta("site_family", defaultSiteFamily).
		WithStartMeta("runner", profile.Runner)
}

File: examples/class-kits/rustspider/ecommerce_detail.rs
use std::sync::Arc;

use rustspider::scrapy::project as projectruntime;
use rustspider::scrapy::{Item, Output, Response, Spider};

use super::ecommerce_profile::{
    best_title, collect_image_links, first_link_with_keywords, first_match, profile_for_family,
    site_family_from_response, text_excerpt, DEFAULT_SITE_FAMILY,
};

pub fn make_ecommerce_detail_spider() -> Spider {
    let profile = profile_for_family(DEFAULT_SITE_FAMILY);
    Spider::new(
        "ecommerce-detail",
        Arc::new(|response: &Response| {
            let family = site_family_from_response(response);
            let current = profile_for_family(&family);
            let links = response.xpath("//a/@href").get_all();
            vec![Output::Item(
                Item::new()
                    .set("kind", "ecommerce_detail")
                    .set("site_family", family)
                    .set("title", best_title(response))
                    .set("url", response.url.clone())
                    .set("item_id", first_match(&response.text, current.item_id_patterns))
                    .set("price", first_match(&response.text, current.price_patterns))
                    .set("shop", first_match(&response.text, current.shop_patterns))
                    .set(
                        "review_count",
                        first_match(&response.text, current.review_count_patterns),
                    )
                    .set(
                        "image_candidates",
                        collect_image_links(&response.url, response.xpath("//img/@src").get_all(), 10),
                    )
                    .set(
                        "review_url",
                        first_link_with_keywords(&response.url, links, current.review_link_keywords),
                    )
                    .set("html_excerpt", text_excerpt(&response.text, 800))
                    .set(
                        "note",
                        "Template for public product detail pages. Extend with site-specific JSON/bootstrap extraction when available.",
                    ),
            )]
        }),
    )
    .add_start_url(profile.detail_url)
    .with_start_meta("site_family", DEFAULT_SITE_FAMILY)
    .with_start_meta("runner", profile.runner)
}

pub fn register() {
    projectruntime::register_spider("ecommerce-detail", make_ecommerce_detail_spider);
}
